dora/core/client/fs/src/main/java/alluxio/client/file/dora/netty/NettyDataWriter.java

```diff
@@ -0,0 +1,549 @@
+    switch (requestType) {
+      case ALLUXIO_BLOCK:
+        return Protocol.RequestType.ALLUXIO_BLOCK;
+      case UFS_FILE:
+        return Protocol.RequestType.UFS_FILE;
+      case UFS_FALLBACK_BLOCK:
+        throw new UnsupportedOperationException("Unsupported request type UFS_FALLBACK_BLOCK. ");
+      default:
+        throw new UnsupportedOperationException("Request type needs to be specified. ");
+    }
+  }
+
+  /**
+   * Creates an instance of {@link NettyDataWriter}.
+   *
+   * @param context    the file system context
+   * @param address    the data server address
+   * @param length     the length of the file to write, set to Long.MAX_VALUE if unknown
+   * @param packetSize the packet size
+   * @param type       type of the write request
+   * @param options    details of the write request which are constant for all requests
+   * @param channel    netty channel
+   */
+  private NettyDataWriter(FileSystemContext context, final WorkerNetAddress address,
+                          long length, long packetSize, RequestType type, OutStreamOptions options,
+                          Channel channel) {
+    mContext = context;
+    mAddress = address;
+    mLength = length;
+    Protocol.WriteRequest.Builder builder =
+        Protocol.WriteRequest.newBuilder().setTier(options.getWriteTier())
+            .setType(getRequestType(type));
+
+    Protocol.CreateUfsFileOptions ufsFileOptions =
+        Protocol.CreateUfsFileOptions.newBuilder().setUfsPath(options.getUfsPath())
+            .setOwner(options.getOwner()).setGroup(options.getGroup())
+            .setMode(options.getMode().toShort()).setMountId(options.getMountId()).build();
+    builder.setCreateUfsFileOptions(ufsFileOptions);
+
+    mPartialRequest = builder.buildPartial();
+    mPacketSize = packetSize;
+    mChannel = channel;
+    mChannel.pipeline().addLast(new PacketWriteResponseHandler());
+  }
+
+  @Override
+  public long pos() {
+    try (LockResource lr = new LockResource(mLock)) {
+      return mPosToQueue;
+    }
+  }
+
+  @Override
+  public Optional<String> getUfsContentHash() {
+    return Optional.empty();
+  }
+
+  /**
+   * Write chunk by sending a write request.
+   * @param bytes the bytes to be written
+   * @param off the offset of the given byte array
+   * @param len the length to be written
+   * @throws IOException
+   */
+  public void writeChunk(byte[] bytes, int off, int len) throws IOException {
+    // TODO(JiamingMai): maybe we can reduce copying here
+    ByteBuf byteBuf = mChannel.alloc().buffer(len);
+    byteBuf.writeBytes(bytes, off, len);
+    writeChunk(byteBuf);
+  }
+
+  @Override
+  public void writeChunk(final ByteBuf buf) throws IOException {
+    final long len;
+    final long offset;
+    try (LockResource lr = new LockResource(mLock)) {
+      Preconditions.checkState(!mClosed && !mEOFSent && !mCancelSent);
+      Preconditions.checkArgument(buf.readableBytes() <= mPacketSize);
```

Comment:
We should divide a large buf into smaller ones here.

This `buf` (originally the `bytes`) is from user applications, and we can not require the minimum length of it.